addons/website/migrations/12.0.1.0/post-migration.py

```diff
@@ -64,44 +66,49 @@ def sync_menu_views_pages_websites(env):
             AsIs(col_name),
         )
     )
-    for agnostic_view_id, specific_view_id in env.cr.fetchall():
-        # Create website-specific page for the copied view
-        agnostic_view = env["ir.ui.view"].browse(agnostic_view_id)
+    for data, specific_view_id in env.cr.fetchall():
+        data = json.loads(data)
         specific_view = env["ir.ui.view"].browse(specific_view_id)
-        agnostic_page = agnostic_view.first_page_id
-        if not agnostic_page:
-            continue
-        specific_page = env["website.page"].search([
-            ("url", "=", agnostic_page.url),
-            ("website_id", "=", specific_view.website_id.id),
-        ])
-        if not specific_page:
-            specific_page = agnostic_page.copy({
-                "is_published": agnostic_page.is_published,
-                "url": agnostic_page.url,
-                "view_id": specific_view_id,
-                "website_id": specific_view.website_id.id,
-            })
-        elif specific_page.view_id == agnostic_view:
-            specific_page.view_id = specific_view_id
-        # Create website-specific menu for the copied page
-        specific_menu = env["website.menu"].search([
-            ("website_id", "=", specific_page.website_id.id),
-            ("url", "=", specific_page.url),
-        ], limit=1)
-        if specific_menu:
-            if specific_menu.page_id:
-                specific_menu.page_id = specific_page
-        else:
-            agnostic_menu = env["website.menu"].search([
-                ("website_id", "=", False),
-                ("url", "=", specific_page.url),
-            ], limit=1)
-            if agnostic_menu:
-                agnostic_menu.copy({
-                    "website_id": specific_page.website_id.id,
-                    "page_id": agnostic_menu.page_id.id and specific_page.id,
+        for page in specific_view.page_ids:
+            menus = env["website.menu"].search([
+                ("id", "child_of", page.website_id.menu_id),
+                ("url", "=", page.url),
+            ])
+            # If menus exist, it means the agnostic view wasn't removed and
+            # they already contain all the needed information, except that they
+            # are linked to the website-agnostic page. Let's fix that.
+            if menus:
+                menus.write({
+                    "page_id": page.id,
                 })
+            # In case the menus disappeared, it's possibly because the
```

Comment:
Can we avoid this removal/creation? All of this can affect subsequent modules that extends that records.

Reply:
It's complex.

The problem is that, if a view is removed because it's website-agnostic and not declared by an addon, the cascade FK constraints attached to it will remove the `website.page` records associated to it, which in turn will remove the `website.menu` records attached to it.

~~What we could do is to empty those fields that would produce the cascading effect and store their values in extra columns in the pre-migration. Then, in the post-migration, delete the records if the pointer records are missing and re-link otherwise.~~

:point_up: While I was writing that I realized that it's absurd. What we're doing here is to actually create new website-specific records from the website-agnostic records that existed before. The website-agnostic record removal is just a side effect of something that is no longer gonna be used. The way odoo would work if the user did manual changes would be that. Thus, the relations within the addon would suffer from that same effect in case of extensions.

I could, however, do a `SELECT * FROM website_menu` in the pre-migration to get all fields, including any extended ones, insert the new values from SQL in the post-migration removing the conflicting fields such as `id`. What do you think?

Reply:
Let's keep it this way for now, as the alternative code is not simpler.